R/rows-db.R

```diff
@@ -391,5 +412,98 @@ sql_rows_delete.tbl_sql <- function(x, y, by, ...) {
     "DELETE FROM ", p$name, "\n",
     "WHERE EXISTS (SELECT * FROM ", p$y_name, " WHERE ", p$compare_qual_qq, ")"
   )
+
   glue::as_glue(sql)
 }
+
+rows_get_or_execute <- function(x, con, sql, returning_cols) {
+  if (is_empty(returning_cols)) {
+    dbExecute(con, sql, immediate = TRUE)
+  } else {
+    returned_rows <- dbGetQuery(con, sql, immediate = TRUE)
+    attr(x, "returned_rows") <- as_tibble(returned_rows)
+  }
+
+  invisible(x)
+}
+
+#' Extract the RETURNING rows
+#'
+#' Extract the RETURNING rows produced by `rows_insert()`, `rows_update()`,
+#' `rows_upsert()`, or `rows_delete().`
+#'
+#' @param x A lazy tbl.
+#'
+#' @return A tibble.
+#'
+#' @export
+get_returned_rows <- function(x) {
+  attr(x, "returned_rows")
+}
+
+#' @param sql SQL code generated with a `sql_rows_()` function.
+#' @param returning_cols A (named) character vector of columns to return.
+#'
+#' @export
+#' @rdname rows-db
+sql_add_returning <- function(sql, x, returning_cols) {
+  if (is_empty(returning_cols)) {
+    return(sql)
+  }
+
+  returning_clause <- sql_returning(x, returning_cols)
+  paste0(sql, if (!is_empty(returning_clause)) paste0("\n", returning_clause))
```

Comment:
I think `sql_returning()` should either return a string or throw an error.

```suggestion
  paste0(sql, "\n", returning_clause)
```